text/0060-bazel.md

```diff
@@ -0,0 +1,137 @@
+---
+feature name: bazel
+start date: 2020-01-16
+rfc pr: (leave this empty)
+related issue: 0060
+---
+
+# Summary
+
+This proposal is for migrating the build system to [Bazel](https://bazel.build/), as it solves
+many if not all of the deficiencies of the current system. This involves adding Bazel resources and rules to the repository to manage
+build and test workflows, and bringing the team up to speed on how to use them.
+
+# Motivation
+
+The CDK currently uses a combination of [Lerna](https://lerna.js.org/), [Yarn Workspaces](https://yarnpkg.com/lang/en/docs/workspaces/), 
+and Bash scripts to orchestrate build and test workflows. This has several advantages for managing a monorepo structure, but can also
+be very slow, difficult to work with and extend, and hard to parallelize amongst the team (ie when one unrelated resource changes, the
+whole project still needs to be rebuilt). 
+
+# Basic Example
+
+Instead of running `bash build.sh` or `npm run build` on a subpackage, an engineer would simply run `bazel build <path-to-resource>` or
+`bazel test <path-to-resource`. Running a specific Bazel rule (for instance orchestrating a release) is also straightforward, with
+`bazel run <path-to-resource>.<action>`.
+
+# Design Summary
+
+In order to setup Bazel, the Bazel-specific rules and resources need to be added to the repository. This means adding `BUILD.bazel` files
+to each subpackage, and then building custom rules around some of the internals of the project (e.g. CFN to TS).
+
+# Detailed Design
+
+The structure of adding Bazel to a project is quite simple. Represented below is a tree of the files
+needed for a standard project
+
+```
+BAZEL.build
+package.json
+|packages|
+  BUILD.bazel
+  |apigateway|
+    BUILD.bazel
+    package.json
```

Comment:
Is there duplication between BUILD.bazel and package.json? For example, where do we specify the module's dependencies all the jsii information?

Reply:
We would do that in the BUILD.bazel, similarly to how package.json handles it. There is work underway right now to pull the dependencies automatically from the package.json, but this would not extend to `jsii` metadata. Personally I think it's ok to migrate `jsii` information out of package.json, especially since it has no bearing on Node distributions (kind of breaking the contract of a Node distributed module).

In the meantime it would be duplicated, which I cop to is increased work, but the dependencies of packages change so infrequently, and if it's work that's missed it is also easily detectable.